deploy/deploy.go

```diff
@@ -215,48 +218,20 @@ type KindSpec struct {
 	Retain                   bool              `yaml:"retain"`
 	Wait                     time.Duration     `yaml:"wait"`
 	Kubecfg                  string            `yaml:"kubecfg"`
-	DeployWithClient         bool              `yaml:"deployWithClient"`
 	GoogleArtifactRegistries []string          `yaml:"googleArtifactRegistries"`
 	ContainerImages          map[string]string `yaml:"containerImages"`
 }
 
 func (k *KindSpec) Deploy(ctx context.Context) error {
-	provider := newProvider()
-	if k.Recycle {
-		clusters, err := provider.List()
-		if err != nil {
-			return err
-		}
-		for _, v := range clusters {
-			if k.Name == v {
-				log.Infof("Recycling existing cluster: %s", v)
-				return nil
-			}
-		}
+	if _, err := execLookPath("kind"); err != nil {
+		return errors.Wrap(err, `install "kind" to deploy`)
 	}
-	if k.DeployWithClient {
-		if len(k.GoogleArtifactRegistries) != 0 {
-			return fmt.Errorf("setting up access to artifact registries %v requires unsetting the deployWithClient field", k.GoogleArtifactRegistries)
-		}
-		if len(k.ContainerImages) != 0 {
-			return fmt.Errorf("loading container images requires unsetting the deployWithClient field")
-		}
-		if err := provider.Create(
-			k.Name,
-			cluster.CreateWithNodeImage(k.Image),
-			cluster.CreateWithRetain(k.Retain),
-			cluster.CreateWithWaitForReady(k.Wait),
-			cluster.CreateWithKubeconfigPath(k.Kubecfg),
-			cluster.CreateWithDisplayUsage(true),
-			cluster.CreateWithDisplaySalutation(true),
-		); err != nil {
-			return errors.Wrap(err, "failed to create cluster using kind client")
+	if k.Recycle {
+		log.Infof("Attempting to recycle existing cluster %q...", k.Name)
+		if err := execer.Exec("kubectl", "cluster-info", "--context", fmt.Sprintf("kind-%s", k.Name)); err == nil {
```

Comment:
should you have a dep check on kubectl? - like is there a guarantee that this order is always maintained?

Reply:
I see two ways of doing it:

1) having each "Spec" list out the dependencies and call checkDependencies for each one before any method

2) having docker and kubectl be dependencies of any deployment and have the spec dependencies only include the remainder of the dependencies (this is the standard that this PR implements)

I can see potentially a 3rd design that steals from (1) and (2) where docker and kubectl are checked for a deployment regardless of the specs and spec dependencies are checked before ever method call, not just deploy.

I think the 3rd option is the best considering these specs should not be called outside of Deployment.Deploy() and are only exported due to yaml marshaling

Reply:
done